### ISSUE TITLE
[$2000] HIGH: Implement Comment Linking

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#20282]
This is a public issue to track implementation of [the Comment Linking feature](https://github.com/Expensify/Expensify/issues/147480) (internal link).

Margelo will work on the App implementation of this project

[Design doc can be found here - also internal](https://docs.google.com/document/d/1v-ZaIRTZL5LIsyPWB0IopBNNgCajf5WC1OA8cVKtd8I/edit#heading=h.ns12dr1f35rx)

# Roll out plan

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0116b1ab0be62d4242</li>
        <li>Upwork Job ID: 1777848479203676160</li>
        <li>Last Price Increase: 2024-04-10</li>
    </ul>
</details>

**Comment by perunt** [comment]
I would like to do this

**Comment by puneetlath** [comment]
@chiragsalian I don't think you need me here so going to un-assign. But feel free to re-assign me if I missed something.

**Comment by perunt** [comment]
@chiragsalian could you apply the `Review` label to this doc?

**Comment by mountiny** [comment]
[Design doc link](https://docs.google.com/document/d/1v-ZaIRTZL5LIsyPWB0IopBNNgCajf5WC1OA8cVKtd8I/edit#heading=h.ns12dr1f35rx) here.

**Comment by sarajojacobson** [comment]
I have read and reviewed this Design Doc.

**Comment by flodnv** [comment]
I have read and reviewed this Design Doc!

**Comment by StefStavri** [comment]
I have read and reviewed this Design Doc!

**Comment by chiragsalian** [comment]
In progress, looks like we got most of the review comments. @robertjchen would you also be able to review the comment linking design doc soon (asking cause you are the only product manager reviewer pending before we can move to the technical section review)

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by chiragsalian** [comment]
Alright looks like we got the required number of high-level reviews. Let's move ahead to the technical review.
@perunt when you get a chance remember to go over all high-level comments and address them.

**Comment by cead22** [comment]
I have read and reviewed this Design Doc!

**Comment by CortneyOfstad** [comment]
I have read and reviewed this Design Doc!

**Comment by johnmlee101** [comment]
I have read and reviewed this Design Doc!

**Comment by laurenreidexpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
I'm going help shepherd this initiative while @chiragsalian's away.

**Comment by roryabraham** [comment]
Still need 3 non-engineering product managers / generalists to review the detailed, and 2 project managers.

**Comment by joelbettner** [comment]
I have read and reviewed this Design Doc!

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by stephanieelliott** [comment]
I have read and reviewed this Design Doc!

**Comment by StefStavri** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
I just updated the rollout plan to include a bit more detail.

At this point, this doc has gone through many reviews and revisions, and we are only short 2 non-engineering generalists, so I'm doing a _last call_ here. This project will go into the implementation phase tomorrow.

**Comment by anmurali** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
Going to move this forward into the implementation phase and start creating issues now.

**Comment by roryabraham** [comment]
Created a project board to start tracking implementation: https://github.com/orgs/Expensify/projects/62/views/1, also created an internal issue to make that project board publicly visible

**Comment by roryabraham** [comment]
Alright, all the issues have been created. Feel free to follow the implementation in the [GitHub Project](https://github.com/orgs/Expensify/projects/62/views/1)

**Comment by JmillsExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
Posted project update in slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1690393798476199

**Comment by HaleyJacobson** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
Weekly update:

- We have a [PR in review](https://github.com/Expensify/Auth/pull/8412) for `GetOlderActions` and `GetNewerActions`
- `previousReportActionID` Onyx migration PR was just merged
- Front-end FlatList bugs are fixed (we still need to finish upgrading react-native-web, which just came off hold)
- I am still working on adding `previousReportActionID` to all reportActions in Auth. It's taken longer than it should have, but should be ready for review by EOD

**Comment by chiragsalian** [comment]
Not overdue, still in progress.

**Comment by roryabraham** [comment]
Hey, sorry my updates here have been lackluster on this issue, but I did drop an update in slack last week [here](https://expensify.slack.com/archives/C035J5C9FAP/p1693329936355999)

**Comment by roryabraham** [comment]
@aldo-expensify raised his hand on https://github.com/Expensify/App/issues/23218, thanks @aldo-expensify! I'm still working on https://github.com/Expensify/Web-Expensify/pull/38662

Also, I'm making this issue a daily since we're shooting to finish this by end-of-offshore

**Comment by roryabraham** [comment]
We had a number of PRs merged for this last week. Still working on the [last backend piece](https://github.com/Expensify/Web-Expensify/pull/38662) then this should be prettymuch all front-end.

**Comment by roryabraham** [comment]
Actively implementing :working:

**Comment by roryabraham** [comment]
`previousReportActionID` isn't really working due to old reportActions being cached in PHP. Created https://github.com/Expensify/Auth/pull/8764 to fix it

**Comment by roryabraham** [comment]
Quick EOW update:

- I still need to re-review https://github.com/Expensify/App/pull/26166, but haven't been urgently prioritizing because there's still an issue with the back-end logic for `previousReportActionID`, which will be needed for the next step which actually enables comment linking
- The [back-end PR](https://github.com/Expensify/Auth/pull/8764) has been ready and approved for 3 days, but has been held up by failing tests in Web-Expensify – some which needed to be temporarily disabled and others which are unrelated and failing on main.

**Comment by roryabraham** [comment]
Has been delayed by my OOO but I just commented on the PR to hopefully help unblock it

**Comment by cristipaval** [comment]
Did Melvin assign me by mistake or I'm missing something

**Comment by roryabraham** [comment]
Going to remove the reviewing label here so it's not greyed out in my K2

**Comment by roryabraham** [comment]
Here's a summary of all the work I'm tracking related to this project:

- https://github.com/Expensify/App/pull/32098
    - Pinged people in slack to try to push to get this ready for review: https://expensify.slack.com/archives/C035J5C9FAP/p1702669706129149 https://expensify.slack.com/archives/C02NK2DQWUX/p1702669521713009
- https://github.com/Expensify/App/pull/32843 (on HOLD for https://github.com/Expensify/App/pull/32098)
- https://github.com/Expensify/App/pull/30269 (on HOLD for both previous issues)
- https://github.com/Expensify/App/issues/23231 (on HOLD for https://github.com/Expensify/App/pull/30269)

Other work which has been discussed but which I'm not actively tracking:

- adding automated UI tests around the comment linking / bidirectional pagination features
- ReportScreen performance audit

**Comment by roryabraham** [comment]
Latest update [here](https://expensify.slack.com/archives/C035J5C9FAP/p1703174259992289), awaiting an update in https://github.com/Expensify/App/pull/32098

**Comment by roryabraham** [comment]
Ok, so @0xmiroslav still hasn't approved https://github.com/Expensify/App/pull/32098 but at this point I think I'm going to just do some thorough review and testing myself

**Comment by roryabraham** [comment]
https://github.com/Expensify/App/pull/32098 has been merged

**Comment by quinthar** [comment]
Closing this for https://github.com/Expensify/App/pull/30269

**Comment by roryabraham** [comment]
@puneetlath I reopened this so that @ishpaul777 can get paid for his role as a C+ for comment linking. I suggest a payment of $1000 with no regression penalties because:

- this was a big, challenging PR
- @ishpaul777 was very diligent in going above-and-beyond to help us resolve any regressions created by the feature

**Comment by roryabraham** [comment]
Hi @ishpaul777, I chatted with @puneetlath and given the difficulty and scope of work, along with all the extra time you spent going above-and-beyond to help deliver the feature, we're willing to authorize $2000 for this issue 👍🏼

**Comment by puneetlath** [comment]
Sent you an offer @ishpaul777: https://www.upwork.com/nx/wm/offer/101826333

Please ping me here when you've accepted.

**Comment by puneetlath** [comment]
Great, paid. Thanks everyone. Exciting to have comment linking out in the wild!
